compiler-core/src/type_/expression.rs

```diff
@@ -707,6 +722,40 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
                 .map_err(|e| e.case_clause_mismatch().into_error(typed_clause.location()))?;
             typed_clauses.push(typed_clause);
         }
+
+        // Because exhaustiveness checking in presence of multiple subjects is similar
+        // to full exhaustiveness checking of tuples or other nested record patterns,
+        // and we currently only do only limited exhaustiveness checking of custom types
+        // at the top level of patterns, only consider case expressions with one subject.
+        if subjects_count == 1 {
+            if let Some(subject_type) = subject_types.get(0) {
+                let value_typ = collapse_links(subject_type.clone());
+                // Currently guards are not considered in exhaustiveness checking,
+                // so we go through all clauses and pluck out only the patterns.
```

Comment:
We want to consider any pattern with a guard to not match anything as the guard could evaluate to false.

Reply:
I think this is a similar situation to nested pattern matches (when there are patterns present for constructor's arguments). Currently this PR doesn't take them into account and the exhaustiveness check is "happy" when a constructor is at least mentioned. So the check defaults to passing.

You're saying I should do the opposite for guards, and default to not passing? And keep the behavior in presence of nested patterns as it is now?

Reply:
Here's an example to show why I think this is similar and I would expect both scenarios to behave similarly, even now without full exhaustiveness checking implemented:
```
pub type MyType {
  BoolBox(Bool)
  OtherConstructor
}

pub fn main1(v) {
  case v {
    BoolBox(b) if b == True -> v
    OtherConstructor -> v
  }
}

pub fn main2(v) {
  // vs
  case v {
    BoolBox(True) -> v
    OtherConstructor -> v
  }
}
```

Reply:
With the previous example forcing the user to add a fallback clause seems natural.
Here's another example to consider:
```
pub type Inner {
  Inner
}

pub type MyType {
  Box(Inner)
  OtherConstructor
}

pub fn main1(v) {
  case v {
    Box(i) if i == Inner -> v
    OtherConstructor -> v
  }
}

// vs
pub fn main2(v) {
  case v {
    Box(Inner) -> v
    OtherConstructor -> v
  }
}
```

Reply:
I think any time we see a guard we should consider it unmatched. Ideally we would do the same if we see any non-variable record field but perhaps that is too much complexity to add here.

Reply:
I'm not super happy about it and I would default to passing the check, instead of nagging the programmer to add fallback clauses when compiler isn't smart enough to see that all possibilities are covered... But maybe if this is supposed to be a check that a programmer can rely on, it will be better in the end, when a proper algorithm is implemented.

I think I can probably make it work the way you want it with a minimal change: when preparing a flattened list of patterns from Case, do not even include patterns from branches that have non empty guards.

Reply:
Updated!

Reply:
I think we should see this as a subset of the future behaviour, so we will want to be consistent with the future handling of guards and with other ML languages. There's no way to know if a guard will be true or not, so we consider the clause not to match anything.

Reply:
Alright! This is in https://github.com/gleam-lang/gleam/pull/1462/commits/9a40cdd45bd1a5424aae071a5b0472f690419103